partiql-spi/src/main/kotlin/org/partiql/spi/function/builtins/FnOverlaps.kt

```diff
@@ -0,0 +1,224 @@
+import org.partiql.spi.function.Function
+import org.partiql.spi.function.Parameter
+import org.partiql.spi.function.RoutineOverloadSignature
+import org.partiql.spi.function.builtins.internal.PErrors
+import org.partiql.spi.function.builtins.internal.PErrors.cardinalityViolationException
+import org.partiql.spi.types.PType
+import org.partiql.spi.utils.FunctionUtils
+import org.partiql.spi.utils.FunctionUtils.isDateTimeType
+import org.partiql.spi.utils.FunctionUtils.isIntervalType
+import org.partiql.spi.value.Datum
+
+/**
+ * SQL Spec 1999: Section 8.12 <overlaps predicate>
+ *
+ * <overlaps predicate> ::=
+ *    <row value expression 1> OVERLAPS <row value expression 2>
+ * <row value expression 1> ::= <row value expression>
+ * <row value expression 2> ::= <row value expression>
+ */
+internal object FnOverlaps : FnOverload() {
+    val name = FunctionUtils.hide("overlaps")
+    override fun getSignature(): RoutineOverloadSignature {
+        return RoutineOverloadSignature(name, listOf(PType.dynamic(), PType.dynamic()))
+    }
+
+    override fun getInstance(args: Array<out PType>): Fn? {
+        val arg0 = args[0]
+        val arg1 = args[1]
+        // Both arguments must be collections (arrays or bags) or unknown (null or missing) types
+        if ((arg0.code() == PType.ARRAY || arg0.code() == PType.BAG || arg0.code() == PType.UNKNOWN) &&
+            (arg1.code() == PType.ARRAY || arg1.code() == PType.BAG || arg1.code() == PType.UNKNOWN)
+        ) {
+            return getCollectionInstance(arg0, arg1)
+        }
+        return null
+    }
+
+    private fun getCollectionInstance(arg0: PType, arg1: PType): Fn {
+        return Function.instance(
+            name = name,
+            returns = PType.bool(),
+            parameters = arrayOf(
+                Parameter("period1", arg0),
+                Parameter("period2", arg1)
+            ),
+            isNullCall = false,
+            isMissingCall = true,
+            invoke = { args ->
+                val period1 = args[0].toList()
+                val period2 = args[1].toList()
+                // Each period must have exactly 2 elements (start, end)
+                if (period1.size != 2) {
+                    throw cardinalityViolationException(period1.size, 2)
+                }
+                if (period2.size != 2) {
+                    throw cardinalityViolationException(period2.size, 2)
+                }
+                // D1: value of the first field of <row value expression 1>
+                val d1 = period1[0]
+                val field1 = period1[1]
+                // D2: value of the first field of <row value expression 2>
+                val d2 = period2[0]
+                val field2 = period2[1]
+                // For the datetime validation:
+                if (!d1.isNull && !d1.isMissing && !isDateTimeType(d1.type)) {
+                    throw PErrors.unexpectedTypeException(d1.type, listOf(PType.date(), PType.time(), PType.timestamp()))
```

Comment:
For the list of possible types, we can also include PType.timez and PType.timestampz. Also for the field1/field2 list.

Let's also define a variable for the d1/d2 possible type list and a variable for field1/field2 such that we don't need as much repetition.